add v3.5 blog post (#145)
spinframework/spin-docs: content/blog/announcing-spin-3-5.md

title = "Announcing Spin v3.5"
date = "2025-11-10T01:27:05.449167Z"
template = "blog_post"
tags = []

[extra]
type = "post"
author = "Michelle Dhanani"

---

> The CNCF Spin project recently released [Spin v3.5.0](https://github.com/spinframework/spin/releases/tag/v3.5.0) which includes support for the WASI Preview 3 (WASIp3) release candidate along with several developer experience enhancements! In this blog post, we’ll dive into a few exciting features of this release and highlight an example of what is enabled with WASIp3 in Spin.

Reviewer: The "p" in "WASIp3" doesn't stand for "Preview". "Point" if one must, but really it's just "p" :)

```suggestion
The CNCF Spin project recently released [Spin v3.5.0](https://github.com/spinframework/spin/releases/tag/v3.5.0) which includes support for a release candidate of the upcoming WASIp3 release, along with several developer experience enhancements! In this blog post, we’ll dive into a few exciting features of this release and highlight an example of what is enabled with WASIp3 in Spin.
```

### WASIp3 and the Spin Experience

WASI Preview 3 isn’t fully stable yet; it’s going through a Release Candidate (RC) stage, and there are likely to be some minor changes before the [final release](https://wasi.dev/roadmap). To ensure a consistent experience for developers, Spin requires you to opt in to using this unstable RC in the HTTP trigger section of your spin.toml file:

```toml
[[trigger.http]]
executor = { type = "wasip3-unstable" }
```

#### A Shiny New Rust SDK

The latest release of the Spin Rust SDK introduces two new WASIp3-related crates:

- **`spin-wasip3-http-macro`** – provides the `#[http_service]` macro, which defines the entry point to your component. Functions annotated with this macro must be marked `async`.
- **`spin-wasip3-http`** – offers the core SDK experience for working with WASIp3 in Spin.

> To enable these crates, make sure to enable the `wasip3-unstable` feature for your `spin-sdk` dependency in your application's `Cargo.toml`.

Author: updated this section to be from dev perspective

The first step of using the SDK is opting into the wasip3-unstable feature as is done on the last line of the `Cargo.toml` file above. Also note Rust 1.90 and the wasm32-wasip2 target are required to build WASIp3 components.

The following src/lib.rs file contains the crux of the component.

```rust
[...]

#[http_service]
async fn handle_concurrent_outbound_http_calls(_req: spin_sdk::http_wasip3::Request) -> anyhow::Result<impl IntoResponse> {

    let spin = pin!(get_content_length("https://spinframework.dev"));
    let book = pin!(get_content_length("https://component-model.bytecodealliance.org/"));

    let (first, len) = match futures::future::select(spin, book).await {
        Either::Left(len) => ("Spin docs", len),
        Either::Right(len) => ("Component model book", len),
    };

    let response = format!("{first} site was first response with content-length {:?}\n", len.0?);

    Ok(response)
}
```

> Here we are creating two async tasks (Futures) that will each fetch the content-length from different URLs. futures::future::select(spin, book).await runs both tasks concurrently and waits until one of them completes. It then matches and returns the site that returned first and the resulting content length.

Reviewer 2: ```suggestion
Here we are creating two async tasks (Futures) that will each fetch the content-length from different URLs. `futures::future::select(spin, book).await` runs both tasks concurrently and waits until one of them completes. It then matches and returns the site that returned first and the resulting content length.
```

Beyond experimental support for WASIp3, the Spin 3.5 release contains the following additional features.

### Static HTTP Responses

You can now easily define static HTTP responses right from within the HTTP trigger section of the spin.toml file rather than a component. The response may contain status code, headers, and body. This is especially great for handling fallbacks and redirects. See example [here](https://spinframework.dev/v3/http-trigger#static-responses-with-the-http-trigger).

```toml
# Example use case: fallback 404 handling
[[trigger.http]]
route = "/..."
static_response = { status_code = 404, body = "not found" }

# Example use case: redirect
[[trigger.http]]
route = "/bob"
static_response = { status_code = 302, headers = { location = "/users/bob" } 
```

### Reference Dependencies by Component ID

Previously, we could specify component dependencies from a registry, a local component on the file system, or from a URL. You can now specify a component from within you application a dependency using the Component ID. See documentation [here](https://spinframework.dev/v3/writing-apps#dependencies-from-a-component-in-the-application).

> ### Open AI Support

Reviewer: ```suggestion
### OpenAI Support
```
